src/database/Database.h

```diff
@@ -196,6 +197,40 @@ class Database : NonMovableOrCopyable
     // Access the optional SOCI connection pool available for worker
     // threads. Throws an error if !canUsePool().
     soci::connection_pool& getPool();
+
+    // Select a set of records using a client-defined query string, then map
+    // each record into an element of a client-defined datatype by applying a
+    // client-defined function (the records are accumulated in the "out"
+    // vector).
+    template <typename T>
+    void selectMap(std::string const& selectStr,
```

Comment:
> Therefore, attempting to use these template functions by including Database.h anywhere except Database.cpp will lead to a link failure. That being said, these functions aren't used outside of Database.cpp right now so you may as well just remove the declarations from the header.

I don't understand what you mean about removing the declarations from the header -- that leads to:

```
database/Database.cpp:436:11: error: out-of-line definition of 'selectMap' does not match any declaration in 'stellar::Database'
Database::selectMap(std::string const& selectStr,
          ^~~~~~~~~
```

Also, I wrote those generically with the intent that we might well be able to start using them to factor out some of the repeated code that currently exists in the *SQL.cpp files, or at least to avoid further repetition in the future.  If we do end up using selectMap\<Foo\> in some source file other than Database.cpp, can't we just make it link by also adding the following to Database.cpp?

```
template void Database::selectMap<Foo>(std::string const& selectStr,
                   std::function<Foo(soci::row const&)> makeT,
                   std::vector<Foo>& out);
```

(See for example https://raymii.org/s/snippets/Cpp_template_definitions_in_a_cpp_file_instead_of_header.html.)

Reply:
You can do that, but it doesn't mean you should. The link you shared starts with the following paragraph:

> In this snippet I'll show you how to place your C++ template definitions in a seperate .cpp file. I'd recommend you to just put template definitions in your header file, or a .hpp file, but if you really want to there is a trick to get them in a seperate .cpp file. The trick is to explicitly instanciate every template you're going to use at the end of the .cpp file. With many different templates and types this becomes cumbersome, but for certain usecases it could be useful.

Basically, if you want to take this approach the burden of proof is on you to justify why it is the right thing to do. I laid out some reasons in a separate comment about why I don't generally support this approach. 

----
My comment was confusing, sorry. You are right that removing them from the header wouldn't work immediately. What I meant was that you could remove them from the header and make them free functions. You'll see that I really like free functions whenever possible. The reason is that it makes it extremely clear that you are respecting encapsulation, and that a function doesn't depend on implementation details. In fact, this change presents a beautiful example in action. Your function `selectMap` is a member function, and therefore proceeds to access private variables such as `mSession`. But `mSession` has a getter `getSession` with a non-trivial implementation! There is a check in there to make sure you are on the main thread. You're bypassing that check, but a free function would not have been able to do so.

To me this is a compelling argument for changing to free functions. But if you aren't convinced there are some other options:
- You can at least make the functions private. As noted in the previous paragraph, they present a foot-gun so they shouldn't be public (it might be safe to bypass the check in your case, but it probably isn't in general).
- You can fix the implementations and move the definitions into the header.

Reply:
> Basically, if you want to take this approach the burden of proof is on you to justify why it is the right thing to do. I laid out some reasons in a separate comment about why I don't generally support this approach.

I was mildly afraid of increasing compilation times and maybe even bloating the object code if the linker isn't paying enough attention (as discussed, for example, [here](https://docs.microsoft.com/en-us/cpp/cpp/source-code-organization-cpp-templates?view=vs-2019).  But I do agree with you that placing the definitions in the header file is stylistically superior, so I'm fine with doing so.

> To me this is a compelling argument for changing to free functions. But if you aren't convinced there are some other options:
> - You can at least make the functions private. As noted in the previous paragraph, they present a foot-gun so they shouldn't be public (it might be safe to bypass the check in your case, but it probably isn't in general).
>- You can fix the implementations and move the definitions into the header.

I am convinced, actually; I think that is a good argument for free rather than member functions.  But I would still prefer them public rather than private, because I did intend them as generic, and, as you show, making them free functions would remove the foot-gun, since it would force them to get at the session through ```Database::getSession()```.  So if they become free functions, then I think that addresses the problems with making them private (at least the one you raised in this conversation).

So how about I take your free-functions suggestion _and_ your one alternative suggestion (making it an adjunct rather than an alternative!) to move the definitions into the header?  As far as I can tell, free functions in the header would satisfy all your arguments in this conversation, as well as my argument that, as generic functions, they should be available to any client of Database, not only to members of Database itself.

That's what I've done in the latest code; please let me know whether that all made sense.